src/openms/source/METADATA/ID/IdentificationDataConverter.cpp

```diff
@@ -0,0 +1,888 @@
+          grouping.groups.insert(new_group);
+        }
+        id_data.registerParentMoleculeGrouping(grouping);
+        sublogger.endProgress();
+      }
+
+      id_data.clearCurrentProcessingStep();
+    }
+    progresslogger.endProgress();
+
+    // PeptideIdentification:
+    Size unknown_query_counter = 1;
+    progresslogger.startProgress(0, peptides.size(),
+                                 "converting peptide identifications");
+    Size peptides_counter = 0;
+    for (const PeptideIdentification& pep : peptides)
+    {
+      peptides_counter++;
+      progresslogger.setProgress(peptides_counter);
+      const String& id = pep.getIdentifier();
+      IdentificationData::ProcessingStepRef step_ref = id_to_step[id];
+      IdentificationData::DataQuery query(""); // fill in "data_id" later
+      if (!step_ref->input_file_refs.empty())
+      {
+        // @TODO: what if there's more than one input file?
+        query.input_file_opt = step_ref->input_file_refs[0];
+      }
+      else
+      {
+        String file = "UNKNOWN_INPUT_FILE_" + id;
+        IdentificationData::InputFileRef file_ref =
+          id_data.registerInputFile(file);
+        query.input_file_opt = file_ref;
+      }
+      query.rt = pep.getRT();
+      query.mz = pep.getMZ();
+      static_cast<MetaInfoInterface&>(query) = pep;
+      if (pep.metaValueExists("spectrum_reference"))
+      {
+        query.data_id = pep.getMetaValue("spectrum_reference");
+        query.removeMetaValue("spectrum_reference");
+      }
+      else
+      {
+        if (pep.hasRT() && pep.hasMZ())
+        {
+          query.data_id = String("RT=") + String(float(query.rt)) + "_MZ=" +
+            String(float(query.mz));
+        }
+        else
+        {
+          query.data_id = "UNKNOWN_QUERY_" + String(unknown_query_counter);
+          ++unknown_query_counter;
+        }
+      }
+      IdentificationData::DataQueryRef query_ref =
+        id_data.registerDataQuery(query);
+
+      IdentificationData::ScoreType score_type(
+        pep.getScoreType(), pep.isHigherScoreBetter(), step_ref->software_ref);
+      IdentificationData::ScoreTypeRef score_ref =
+        id_data.registerScoreType(score_type);
+
+      // PeptideHit:
+      for (const PeptideHit& hit : pep.getHits())
+      {
+        if (hit.getSequence().empty()) continue;
+        IdentificationData::IdentifiedPeptide peptide(hit.getSequence());
+        peptide.processing_step_refs.push_back(step_ref);
+        for (const PeptideEvidence& evidence : hit.getPeptideEvidences())
+        {
+          const String& accession = evidence.getProteinAccession();
+          if (accession.empty()) continue;
+          IdentificationData::ParentMolecule parent(accession);
+          parent.processing_step_refs.push_back(step_ref);
+          // this will merge information if the protein already exists:
+          IdentificationData::ParentMoleculeRef parent_ref =
+            id_data.registerParentMolecule(parent);
```

Comment:
do I understand this correctly that you store a list of proteins in `id_data` and here you just look up whether that protein already exists in the list, and if so you always reference the same molecule? Is that also how the memory savings are achieved?

Reply:
Yes.